src/device/socket/vsock.rs

```diff
@@ -398,28 +382,34 @@ impl<H: Hal, T: Transport> VirtIOSocket<H, T> {
         // buffer back to the RX queue then we don't access it again until next time it is popped.
         let header = unsafe {
             let buffer = Self::as_mut_sub_rx_buffer(self.rx_buf_dma.raw_slice(), token.into())?;
-            let _len = self.rx.pop_used(token, &[], &mut [buffer])?;
+            let packet_len = self.rx.pop_used(token, &[], &mut [buffer])?;
 
             // Read the header and body from the buffer. Don't check the result yet, because we need
             // to add the buffer back to the queue either way.
-            let header_result = read_header_and_body(buffer, body);
+            let header_result = read_header_and_body(buffer, body, packet_len);
 
             // Add the buffer back to the RX queue.
             let new_token = self.rx.add(&[], &mut [buffer])?;
             // If the RX buffer somehow gets assigned a different token, then our safety assumptions
             // are broken and we can't safely continue to do anything with the device.
-            assert_eq!(new_token, token);
-
+            if token != new_token {
+                return Err(SocketError::RecycledWrongBuffer.into());
+            }
             header_result
         }?;
 
+        // Notifies the RX queue after adding back the buffer.
+        if self.rx.should_notify() {
```

Comment:
I'd rather keep this outside of the loop for efficiency, so that it only gets called once rather than multiple times if we loop through several entries in the queue. Sending the notification will result in a context switch to the host VMM, so it's a relatively expensive operation.